src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,270 @@
+
+/**
+ * A stack of function IDs.
+ */
 type Stack = Array<number>;
+
+/**
+ * A function which converts entry into one or more samples, then
+ * appends those sample(s) to samples.
+ */
+type AppendEntryToSamples =
+    (entry: Entry, samples: perftools.profiles.Sample[]) => void;
+
+/**
+ * Profile node and stack trace to that node.
+ */
 interface Entry {
-  node: TimeProfileNode;
+  node: ProfileNode;
   stack: Stack;
 }
 
 /**
- * Converts v8 Profile into profile with profile format used by Stackdriver
- * Profiler.
- *
- * @param prof - profile to be converted.
- * @param intervalMicros - average time (microseconds) between samples.
+ * Heap profile node and stack trace to that node.
  */
-export function serializeTimeProfile(
-    prof: TimeProfile, intervalMicros: number) {
-  let samples: Array<perftools.profiles.Sample> = [];
-  let locations: Array<perftools.profiles.Location> = [];
-  let functions: Array<perftools.profiles.Function> = [];
-  let locationMap: Map<number, perftools.profiles.Location> = new Map();
-  let functionMap: Map<number, perftools.profiles.Function> = new Map();
-  let strings = [''];
-
-  let sampleValueType = createSampleValueType();
-  let timeValueType = createTimeValueType();
+interface AllocationEntry {
+  node: AllocationProfileNode;
+  stack: Stack;
+}
 
-  serializeNode(prof.topDownRoot);
+/**
+ * Time profile node and stack trace to that node.
+ */
+interface TimeEntry {
+  node: TimeProfileNode;
+  stack: Stack;
+}
 
-  return {
-    sampleType: [sampleValueType, timeValueType],
-    sample: samples,
-    location: locations,
-    function: functions,
-    stringTable: strings,
-    // opt drop_frames
-    // opt keep_frames
-    timeNanos: 1000 * 1000 * prof.endTime,                         // Nanos
-    durationNanos: 1000 * 1000 * (prof.endTime - prof.startTime),  // Nanos
+/**
+ * Used to build string table and access strings and their ids within the table
+ * when serializing a profile.
+ */
+class StringTable {
+  strings: string[];
+  stringsMap: Map<string, number>;
 
-    periodType: timeValueType,
-    period: intervalMicros
-  };
+  constructor() {
+    this.strings = [];
+    this.stringsMap = new Map<string, number>();
+    this.getIndexOrAdd('');
+  }
 
   /**
-   * Adds samples from a node and it's children to the fields tracking
-   * profile serialization.
-   *
-   * @param node - the node which is serialized
-   * @param stack - the stack trace to the current node.
+   * Returns index of str within the table. Also adds str to string table if
+   * str is not in the table already.
    */
-  function serializeNode(root: TimeProfileNode) {
-    let entries: Entry[] = [];
-    // don't include root node in serialized profile, start with it's children.
-    for (let child of root.children) {
-      entries.push({node: child, stack: []});
+  getIndexOrAdd(str: string): number {
+    let loc = this.stringsMap.get(str);
```

Comment:
Nit: loc -> idx since that's how you call that in the function doc.

Reply:
Done